### PR TITLE
Refactor: extract shared post-training hooks and update SFT implementation

### PR DESCRIPTION
**Pull request by igorts-git**
# Description

Generalize the SFT hooks such that they can be used in DPO as well.

We move the shared code into `src/maxtext/trainers/post_train/hooks.py` while the SFT-specific overrides reside in the SFT sub-directory.

This PR is a pre-requisite for #3668 (DPO implementation)

# Tests

Test run of a local SFT training.
Ran the unit tests, including logits comparison test.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by shralex — APPROVED** [review]

File: tests/post_training/unit/hooks_test.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
+# Copyright 2023–2025 Google LLC
```
Comment (SurbhiJainUSC) [inline tests/post_training/unit/hooks_test.py]: nit: `2023-2026`

**Review by SurbhiJainUSC — APPROVED** [review]

File: tests/post_training/unit/hooks_test.py (hunk 1 above)
Comment (igorts-git) [reply tests/post_training/unit/hooks_test.py]: done

**Review by NicoGrande — APPROVED** [review]
